Refactored online installer/uninstaller

miniche opened the pull request:
Before the version 0.1.0, I've already created a "online installer/uninstaller".
With this feature, anyone can install CentralReport on his host with a simple command in a terminal, without downloading the project before.

In the idea, you open the terminal, and you copy/past the following command:

``` bash
curl http://scripts.centralreport.net/install | bash
```

(This URL doesn't exist yet, but it will be the "tools/online_installer.sh" script)

This script...
- Checks the current OS
- Checks if Python is available
- Downloads a "full package"
- Unarchives the full package
- Updates the chmod on install.sh script (chmod +x install.sh)
- Executes the install.sh script
- Deletes the unarchived package
- Deletes the package downloaded

The full package must be generated before with the "tools/packager.sh" script.
This packager generate an archive (.tar.gz) with the complete CentralReport script (as you download it manually from Github in the master branch).
It must be available from the centralreport server, and the URL fixed in online installer and uninstaller.

With this feature, installing or deleting CentralReport from any host connected to Internet will be easy.

(PS: the same solution exists also for the uninstaller)

miniche committed:
Updated "packager" with new bash functions

miniche committed:
Updated online installer/uninstaller

New URLs and little fixes

miniche commented:
We need three URLs for this feature:
- A URL where the full package is available
- A URL for the online_installer.sh script
- A URL for the online_uninstaller.sh script

I suggest:
- http://scripts.centralreport.net/install
- http://scripts.centralreport.net/uninstall
- http://files.centralreport.net/package.tar.gz

But, perhaps have you better ideas? :smile:

Ninir commented on tools/online_installer.sh:
```diff
@@ -7,20 +7,20 @@
 OS_DEBIAN="Debian"
 
-echo -e "\n\nWelcome to CentralReport one-line installer!"
+echo -e "\n\nWelcome to CentralReport online installer!"
```
Not sure about the online part here. When referring to online, I think about an online installer, which is not really the case here. This is a command which downloads a package from the internet, right?

miniche replied:
In this case... What is your definition of a "online installer"?

Yes, this script downloads a package, but do a little more than this. I've detailled all steps in the PR description ("This script..." part)

Ninir replied:
I would say "online" installer for something that is done on a web page, online.

THere might be other words to go with here... let's find them! :smile:

miniche replied:
Mmh ok, I understand what you mean.
Before, I'd used the "one-line" composed word because with this functionality, you just write one line in you terminal, and all the (hard) work is done.

If you find a better wording, I take! :stuck_out_tongue_winking_eye:

Ninir replied:
A perfect example could be: CloudShare or WindowsAzure, you can install/configure your  VM online. Is it more accurate? :/
Don't have better examples atm :/

miniche replied:
I think it will be hard to find the _good_ word :smile:

Ninir replied:
Let's keep it for now, you got me ^^ (don't have a better proposition ^^)
However, **please add a todo** as a comment so that we know we should find a better way (maybe when we'll have a native english speaker :P)

Ninir replied:
ping @miniche

miniche replied:
Oh... I seems that I forgot to do that...

Ninir commented on tools/online_uninstaller.sh:
```diff
@@ -7,19 +7,21 @@
 # ------------------------------------------------------------
 
-# This script will download latest CentralReport version, and begin uninstall on current host.
+# This script will download latest CentralReport version, and uninstall CR from this host
```
`and uninstall the previous CR version from this host`

Btw, you don't need the coma after version

Ninir replied:
If it's an uninstaller... why should you download the latest CentralReport version?
(Didn't check how it works:) Is it meant to be used in another script?

miniche replied:
Because you're doing more than a simple folder deletion (root access, unregistering init script, etc.), and all this steps are executed by bash functions. 

We can also make a "lite" package, with just bash scripts, without the python part (just the "bash" folder and the uninstall.sh script). That could be great, no?

Ninir commented on tools/online_installer.sh:
```diff
@@ -7,20 +7,20 @@
 # ------------------------------------------------------------
 
-# This script will download latest CentralReport version, and begin installation.
+# This script will download latest CentralReport version, and begin the installation.
```
You don't need the coma after version

Ninir commented on tools/online_uninstaller.sh:
```diff
@@ -40,7 +42,16 @@ else
 fi
 
-# Downloading full package...
+# Testing if Python is available on this host
```
`Checking whether Python is available on this host`

Ninir commented on tools/online_uninstaller.sh:
```diff
@@ -40,7 +42,16 @@ else
 
-# Downloading full package...
+# Testing if Python is available on this host
+echo -e "\nTesting Python version..."
```
According to your comment, this is not the appropriate sentence to print IMO.
What about `Checking Python existence...`?

Ninir commented on tools/packager.sh:
```diff
@@ -10,54 +10,70 @@
-sudo rm -R CentralReportIndev/.DS_Store
+logConsole "Removing useless dot files..."
+sudo rm -R CentralReportPackage/.git 2>/dev/null
+sudo rm -R CentralReportPackage/.idea 2>/dev/null
```
This should not even be there in fact. It should be removed from the Github repo

miniche replied:
Yes, but the packager take the current folder on your "dev" computer, not from the repository. And in this case, we have this `.idea` folder.
Another solution can be delete all dot files (.*).

Ninir replied:
There are 2 ways to proceed here:
- Making a whitelist, blacklisting what you don't want
- Making a blacklist (saying that every files/dir should not be included), whitelisting what you want to add.

So instead of taring the whole dir, I would go for a selection in files/dir. 

What do you think?

miniche replied:
Good idea for a blacklist. It's more flexible than a whitelist in our case.

We need two blacklists:
- For the "installer" package (with all scripts included)
- For the "uninstaller" package (with just files required to perform the uninstallation)

I will try to implement this before the end of the week.

Ninir replied:
Thank you, you're the man :P

miniche replied:
Or perhaps a whitelist... I'm thinking about that.
The whitelist can contain files or folders, for more flexibilities, and it may be easier to maintain:
- It's easier to detect missing files/folders with tests instead of unused files/folders
- With a blacklist, it will be hard to predict all unwanted files (.idea, .eclipse, .DS_store, etc.) for all developers

What do you think about that?

miniche committed:
Spelling fixes

miniche committed:
Merge remote-tracking branch 'origin/develop' into feature-online-installer

miniche committed:
Added whitelists and blacklists for the packager (installer/uninstaller)

- Two packages are now generated: a package for the installer and a package for the uninstaller.
- The whitelist is read before the blacklist. In fact, all items in the blacklist are deleted in the files/folders copied by the whitelist.

The lists can contain:
- Files
- Folders (all files and sub-folders in the folder will be copied)

The blacklists can also contain items starting by *. A search in all directories will be performed and all corresponding elements will be deleted (e.g. for all *.pyc files)

miniche committed:
Moved "packager.sh" script to "make_packages.sh" in the packager folder

miniche commented:
We have now two packages: the "installer" and the "uninstaller".
Each package contains only useful files for their respective missions.

After a long meditation, I've chosen to create a whitelist and a blacklist for both.
The whitelist is read first. It contains a list of folders or files needed. Each item is written on a new line, and they will be copied on a temporary folder (in /tmp).
The blacklist is read after the whitelist. All unneeded items will be deleted from the temporary folder.

Why the both?
In fact, the whitelist is useful to describe all items needed for the packages. For example, we want to keep the folder "centralreport/" and its content, recursively. But, some folders, we can found ".pyc" file (files compiled by Python: http://docs.python.org/release/1.5.1p1/tut/node43.html). These files can't be included in the final packages.
The blacklist can contain files, folders and bash regex (starting by a star (*)). With the rule *.pyc, all files with the pyc extension will be deleted from the temporary folder, used for the package generation.

Finally, the installer package size is 971 Ko, and the uninstaller size is only 13 Ko.

Guimove commented on tools/packager/uninstaller_blacklist.txt:
```diff
@@ -0,0 +1,12 @@
+# The blacklist is read after the whitelist.
+
+*.pyc
+*.DS_Store
```
The .idea folder is not present in this blasklist ?

miniche replied:
No, because the .idea is not copied by the whitelist.
The blacklist is only useful to delete unneeded files/folders from the temporary folder, generated reading the whitelist.

Guimove commented:
For the rest that's ok

miniche commented on tools/packager/installer_blacklist.txt:
```diff
@@ -0,0 +1,12 @@
+# ------------------------------------------------------------
+# CentralReport Unix/Linux installer whitelist
```
To me: It's the blacklist, not the whitelist!

miniche commented on tools/packager/uninstaller_blacklist.txt:
```diff
@@ -0,0 +1,12 @@
+# ------------------------------------------------------------
+# CentralReport Unix/Linux installer whitelist
```
Wrong header: it's the uninstaller blacklist.

miniche commented on tools/packager/uninstaller_whitelist.txt:
```diff
@@ -0,0 +1,17 @@
+# ------------------------------------------------------------
+# CentralReport Unix/Linux installer whitelist
```
Wrong header: it's the uninstaller whitelist.

Ninir commented:
@miniche Commenting your own Pull Request make us think you are schizophrenic... ^^

miniche commented:
It's so hard to be agree with oneself... :D
I've noticed this just before a meeting, it's just to avoid forgetting :)

Guimove commented:
This man is always on a meeting... :trollface:

miniche committed:
Fixed header in whitelist/blacklist for the installer and the uninstaller

miniche commented:
@Ninir, what do you think about the lists? And you can do a review if you want :P

miniche committed:
Installer/Uninstaller: Added more checks and cleaned comments

miniche committed:
The packager have now its own functions library

It's cleaner and we can now detect any error during the process, and cleaning the "/tmp" folder at the end.

miniche committed:
Improved the online installer/uninstaller: comments and variables.

miniche committed:
Added online scripts in the package directory

It will be more easy to find them on copy them on a remote server after the package creation.

miniche commented:
I have made some improvements for this feature.
Now, the "packager" have its own functions: it's cleaner, and we can catch more exceptions and errors.
The "online" scripts are now copied in the final folder (in /tmp/cr-package for now) to put them on the CRO server.

You can test this executing the following command:

```
curl charles-emmanuel.me/cr/install | bash
```

(and the uninstaller is available with `curl charles-emmanuel.me/cr/uninstall | bash`)

I'm waiting for your reviews and suggestions :)

miniche commented:
Hey guys, don't forget to review this pull request :)

Guimove commented:
In Debian, curl was not installed, I had to install it manually.

In all my VMs, I get this issue : 

```
curl: (6) Couldn't resolve host 'charles-emmanuel.me'
```

But a ping works :angry:

miniche commented:
Right!

This command may work for Debian/Ubuntu:

```
wget -qO- http://charles-emmanuel.me/cr/install | bash
```

I forgot to add this command in my previous comment...
Can you confirm this for me?

Guimove commented:
Nop not working too

![capture_d_cran_27_05_13_23_50](https://f.cloud.github.com/assets/1172067/569864/836dbc1c-c717-11e2-8617-b0d505778ca5.png)

miniche commented:
I see that you're using zsh, I have only tried with bash. I will do more tests on my VM with zsh in this case.

Guimove commented:
Are you sure about the URL ?

![capture_d_cran_27_05_13_23_51](https://f.cloud.github.com/assets/1172067/569870/b654b202-c717-11e2-8f51-c8fe1f231956.png)

miniche commented:
My bad... I modified the Apache configuration on my server during the last week... Shame on me.
I will fix it!

Guimove commented:
:scream: :rage: 

I'll check tomorrow now.

We have to had the support for CentOS too now following thie PR #58

miniche commented:
:worried:

Fixed!
![capture decran 2013-05-27 a 23 58 46](https://f.cloud.github.com/assets/1638158/569887/b8e30356-c718-11e2-85fc-23fd09c77ada.png)

miniche commented:
> I'll check tomorrow now.
> We have to had the support for CentOS too now following thie PR #58

:+1: 
We can finish the PR #58, merging it in develop and update this pull request after to add the compatibility with CentOS. Do you agree?

Guimove commented:
Ok for that but after the merge of the CentOS branch, we'll have to integrate this OS in all futur devs.

Guimove commented:
Working fine for me now

miniche commented:
Cool :)
And what do you think about this feature? These main advantages, in my opinion: it's cross-platforms and it's very simple. No package to be downloaded manually, no complex operations...

The ultimate work should be integrate CR in the official linux repositories (Debian, Ubuntu...). We can install CR with a simple "apt-get install centralreport" :D

Ninir commented:
Please make the package :)

Guimove commented:
We can make us own debain repository...let's check how to do this.

Here is some exemple : 
Package : http://alp.developpez.com/tutoriels/debian/creer-paquet/
Repo : http://blog.jonliv.es/2011/04/26/creating-your-own-signed-apt-repository-and-debian-packages/

miniche commented:
@Ninir: I've made the package on my own server for testing purposes.
If we want to generate it on our server, we must be agree about the URLs (Please see: https://github.com/CentralReport/CentralReport/pull/70#issuecomment-17134733)

@Guimove: Yeah, but in this case, we must update our own repositories list in the configuration file on Debian, doesn't it? I think it's better to find a solution to integrate the CR debian package in the official repositories.

Ninir commented:
> We need three URLs for this feature:
> 
> A URL where the full package is available
> A URL for the online_installer.sh script
> A URL for the online_uninstaller.sh script
> I suggest:
> 
> http://scripts.centralreport.net/install
> http://scripts.centralreport.net/uninstall
> http://files.centralreport.net/package.tar.gz
> But, perhaps have you better ideas? 

Seems good :)

miniche commented:
This PR will be updated with the work on #58 (CentOS distribution support) in the next days.

miniche committed:
Merge remote-tracking branch 'origin/develop' into feature-online-installer

miniche committed:
Added CentOS support in the online installer/uninstaller

miniche committed:
Updated NEWS file with the online installer/uninstaller feature

miniche commented:
I've generated the packages and the scripts from this current branch.

You can now make tests on CentOS and OS X:

```
curl http://scripts.centralreport.net/install | bash
```

On Debian/Ubuntu (they doesn't have "curl" installed by default):

```
wget -qO- http://scripts.centralreport.net/install | bash
```

To uninstall CentralReport, just replace "install" by..."uninstall" in the URLs.

Please give me your feelings :)

Guimove commented:
Ok for me in all my VMs

miniche commented:
Great :)
@Guimove, do you agree with the proposed URLs?

Ninir commented:
@miniche "do" you agree*

Gauthier Wallet
Envoyé depuis mon iPhone.

Le 12 juin 2013 à 21:19, Charles-Emmanuel notifications@github.com a écrit :

> Great :)
> @Guimove, are you agree with the proposed URLs?
> 
> —
> Reply to this email directly or view it on GitHub.

miniche commented:
This is what I wrote... What is the problem? :D
(Thanks, it was a ugly mistake...)

Guimove commented:
IMO their is to much subdomains, why not simply using one only entry point for downloads ?
Like : 
''''
Http://dl.central...
''''

miniche commented:
Yes, it may be better to keep one subdomain. Why not for "dl.centralreport.net".

Ninir commented:
What about "static.centralreport.net"? many sites are using this approach for their static files (images, scripts, ...)

miniche commented:
In fact, I doesn't have a better idea. We can have a dedicated subdomain for the CentralReport packages (installer/uninstaller, Unix/Linux and Windows installer after), or having a more generic subdomain.

Wouldn't it be better to have a subdomain for this functionality to keep the files organized and a simple URL, easy to remember?

Just some ideas:
- curl dl.centralreport.net/install
- curl linux.centralreport.net/install
- curl bash.centralreport.net/install
- curl NoMoreIdea.centralreport.net/install

miniche commented:
Ping @Guimove @Ninir about https://github.com/CentralReport/CentralReport/pull/70#issuecomment-19427440

Guimove commented:
I prefere the "static" or "dl" only because we don't have to create and manage a lot of subdomains.
Let's vote ?

Ninir commented on NEWS:
```diff
@@ -19,6 +19,12 @@ Version Alpha 0.3.0 (Not released yet)
           All confirmations are validated by default.
           Useful to perform tests or for deployments.
 
+    - Feature #70: New online installer and online uninstaller
```
`New online installer and uninstaller` is enough :)

Ninir commented on NEWS:
```diff
@@ -19,6 +19,12 @@ Version Alpha 0.3.0 (Not released yet)
           Useful to perform tests or for deployments.
 
+    - Feature #70: New online installer and online uninstaller
+        * Install and uninstall CentralReport with a single line in a terminal.
```
Install and/or uninstall

miniche replied:
Why `and/or`? I'm not sure that is a better formulation in this case (in English as well in French).

"You can install and uninstall CR with a sigle line in a terminal", in other words "You can do the both with your terminal". If you want to insist on the two separate scripts, I think it's better to use "or". 

(And for the story, the "et/ou" in French is not recommended, except in the familiar language and in newspapers: http://www.druide.com/enquetes/la-pseudo-conjonction-etou
In more formal writing, it must be avoided.

In English, it seems to be the same:
http://billbennett.co.nz/2009/09/26/writing-avoid-andor-slashes/
http://grammar.ccc.commnet.edu/grammar/plague.htm
http://grammar.quickanddirtytips.com/and-or-singular-plural.aspx)

Ninir commented on NEWS:
```diff
@@ -19,6 +19,12 @@ Version Alpha 0.3.0 (Not released yet)
 
+    - Feature #70: New online installer and online uninstaller
+        * Install and uninstall CentralReport with a single line in a terminal.
+          One remote script is available for the installer and another for the
```
and another one

Ninir commented on NEWS:
```diff
@@ -19,6 +19,12 @@ Version Alpha 0.3.0 (Not released yet)
+        * Install and uninstall CentralReport with a single line in a terminal.
+          One remote script is available for the installer and another for the
+          uninstaller. They download the corresponding package automatically
+          and they start the processes.
```
and then start processes. (we don't need the "the" word as you don't provide the process list)

Ninir commented on tools/online_uninstaller.sh:
```diff
@@ -7,18 +7,22 @@
 # ------------------------------------------------------------
 
-# This script will download latest CentralReport version, and begin uninstall on current host.
+# This script will download latest CentralReport version and uninstall CR from this host
```
Does it really download the latest CentralReport version or does it download only the uninstaller? (Looking ath the URL_CR, it downloads only the latest uninstaller).

What about `This script will download the latest CentralReport uninstaller and process the uninstall`?

miniche replied:
It's true :)

Ninir commented on tools/online_installer.sh:
```diff
@@ -7,20 +7,22 @@
 # ------------------------------------------------------------
 
-# This script will download latest CentralReport version, and begin installation.
+# This script will download latest CentralReport version and begin the installation.
```
See https://github.com/CentralReport/CentralReport/pull/70/files#r4719125 for the same idea

miniche replied:
In this case, this script downloads the "installer package", which contains the lastest CentralReport version.

miniche committed:
Spelling fixes in the online installer part

miniche commented:
After our discussion this afternoon about that, the three URL will be:
- static.centralreport.net/installer
- static.centralreport.net/uninstaller
- static.centralreport.net/cr_installer.tar.gz
- static.centralreport.net/cr_uninstaller.tar.gz

Do you agree?

miniche committed:
Updated URL for the online installer and uninstaller

Guimove commented:
Go for that

miniche commented:
Scripts and packages have been generated and they are available here: http://static.centralreport.net/.
I will update the documentation tonight.

(Packages have been created with the current version in the master branch, the 0.2.0)

miniche commented:
Documentation updated: http://docs.centralreport.net/unix-agent/install

Ready to be merged.

Guimove commented:
Ok for me

Ninir commented:
:+1:

miniche commented:
I have just a little problem with our online server tonight. With the command `curl http://static.centralreport.net/installer`, our web server returns "HTTP/1.1 406 Not Acceptable".
(But it works fine with wget...)

I'm trying to fix it.

Ninir commented:
Try using `-O` as in `curl -O http://static.centralreport.net/installer`. It will then download the file locally using the remote name.

Guimove commented:
or like in the doc : 

```
curl http://static.centralreport.net/installer | bash
```

miniche commented:
@Ninir It doesn't work for me:
![capture](https://f.cloud.github.com/assets/1638158/708893/f067f1c4-de56-11e2-91fc-b7a0d1cc8c3e.png)

@Guimove I have the same error in this case:
![capture2](https://f.cloud.github.com/assets/1638158/708901/454ccdb8-de57-11e2-95be-4fae6ac3cb01.png)

And when we had tried this functionality (https://github.com/CentralReport/CentralReport/pull/70#issuecomment-19297597) with other URL, everything worked well...

Guimove commented:
After some search, it looks like apache (or the web serveur used by CRO) is blocking cURL form accessing static file.

You can bypass this security faking your user agent (crappy solution not usable) : 

```
curl http://static.centralreport.net/installer -A "Mozilla"| bash
```

or we should look with us hosting or mod.security configuration

miniche commented:
It seems that we have a problem with the actual configuration of our web server.
We are investigating about this.

Guimove commented:
IMO you can merge this branch as the problem is not in the code but it's only a server configuration mistak

Ninir commented:
Merging a branch means that the work behind will be available to everyone once it's in master :/

The issue spotted is about curl itself: last checks proved that wget was also issuing changing the user-agent to be the same used by curl.

We should really find the issue and fix it as it is blocking potentiel users IMO.

What do you think guys?

miniche commented:
I think we must wait to have a fully functional solution before merging this branch into develop.

IMO, our priorities are:
- Finish the work on the branch feature-webservices (PR https://github.com/CentralReport/CentralReport/pull/52)
- Finish the work on the branch feature-manager (PR https://github.com/CentralReport/CentralReport/pull/82)
- Solve this problem
- Release the version 0.3.0

We must speak about this issue and our "online architecture" together :wink:

miniche commented:
Guys,

Packages are now available at `http://static.centralreport.net`.
@Guimove, can you try on your VMs with **curl** following the documentation? (http://docs.centralreport.net/unix-agent/install)

Packages will be updated when the version 0.3.0 will be out.

Guimove commented:
That's running well in my VMs. :+1:

miniche commented:
Great :)
@Ninir, you must say "Ok". Just "Ok" :D

Ninir commented on NEWS:
```diff
@@ -19,6 +19,12 @@ Version Alpha 0.3.0 (Not released yet)
+        * Install and uninstall CentralReport with a single line in a terminal.
+          One remote script is available for the installer and another one for
+          the uninstaller. They download the corresponding package automatically
+          and they start processes.
```
`they start processing`. Btw, what about "They are in charge to automatically download the related package and they are handling the magic for you." :D

miniche replied:
Not serious enough. I love :D

Ninir commented on tools/online_installer.sh:
```diff
@@ -7,20 +7,22 @@
 # ------------------------------------------------------------
 
-# This script will download latest CentralReport version, and begin installation.
+# This script will download the latest CentralReport installer and process the install
```
`and processes`

miniche replied:
I must be more careful about your proposals :P (https://github.com/CentralReport/CentralReport/pull/70#discussion_r4719125)

Ninir commented on tools/online_uninstaller.sh:
```diff
@@ -7,18 +7,22 @@
 # ------------------------------------------------------------
 
-# This script will download latest CentralReport version, and begin uninstall on current host.
+# This script will download the latest CentralReport uninstaller and process the uninstall
```
`processes`*

Ninir commented on tools/packager/functions.inc.sh:
```diff
@@ -0,0 +1,217 @@
+                return 2
+            fi
+        fi
+    done
```
To remove indentation to this part (making it more readable), I would suggest to do this:

``` bash
for line in $(cat "$1");
do
    if [[ ${line} == "#"* ]] || [[ ${line} == ";"* ]] ; then
        continue
    fi

    if [ -d ${line} ]; then
        sudo cp -R ${line} "${2}${line}"
        RETURN_CODE="$?"

        if [ "$?" -ne "0" ]; then
            logError "Error copying ${line}!"
            return 1
        fi

    elif [ -f ${line} ]; then
        sudo cp ${line} "${2}${line}"
        RETURN_CODE="$?"

        if [ "$?" -ne "0" ]; then
            logError "Error copying ${line}!"
            return 1
        fi

    else
        logError "Missing file or directory: ${line}"
        return 2
    fi
done
```

Do you get the idea? What do you think?

miniche replied:
Yes, it's clearer!

miniche replied:
Modified in c697fe0.

Ninir commented on tools/packager/functions.inc.sh:
```diff
@@ -0,0 +1,217 @@
+            if [ -d ${line} ]; then
+                sudo cp -R ${line} "${2}${line}"
+                RETURN_CODE="$?"
+                if [ "$?" -ne "0" ]; then
```
Instead of `"$?"`, why not checking `$RETURN_CODE`?

(ima guessing it contains $?)

miniche replied:
Why? You ask me why? No explanation. Bad developers, perhaps :D

miniche committed:
Minor fixes and improvements

miniche commented:
Anything else? :)

Ninir commented:
Seems ok :)

miniche commented:
Packages are already available at http://static.centralreport.net. They are based on the version 0.2.0.
They will be updated when we will release the version 0.3.0.
